src/core/common/linked_list.hpp

```diff
@@ -325,29 +357,99 @@ template <typename Type> class LinkedList
+     * in the list. To check that an entry matches the given indicator, the `Matches()` method is invoked on each
+     * `Type` entry in the list. The `Matches()` method should be provided by `Type` class accordingly:
+     *
+     *     bool Type::Matches(const Indicator &aIndicator) const
+     *
+     * @param[in]  aIndicator  An indicator to match with entries in the list..
+     * @param[out] aPrevEntry  A pointer to output previous entry on success (when a match is found in the list).
+     *                         @p aPrevEntry is set to nullptr if the matching entry is head of the list. Otherwise it
+     *                         is updated to point to previous entry before the matching entry in the list.
+     *
+     * @retval OT_ERROR_NONE       A matching entry was found in the list and @p aPrevEntry was updated successfully.
+     * @retval OT_ERROR_NOT_FOUND  No matching entry was not found in the list.
+     *
+     */
+    template <typename Inidcator> const Type *FindMatching(const Inidcator &aIndicator, const Type *&aPrevEntry) const
+    {
+        const Type *entry;
+
+        aPrevEntry = nullptr;
+
+        for (entry = mHead; entry != nullptr; aPrevEntry = entry, entry = entry->GetNext())
         {
-            for (Type *cur = mHead; cur->GetNext() != nullptr; cur = cur->GetNext())
+            if (entry->Matches(aIndicator))
             {
-                if (cur->GetNext() == &aEntry)
-                {
-                    aPrevEntry = cur;
-                    error      = OT_ERROR_NONE;
-                    break;
-                }
+                break;
             }
         }
 
-        return error;
+        return entry;
+    }
+
+    /**
+     * This template method searches within the linked list to find an entry matching a given indicator, and if found
+     * returns a pointer to its previous entry in the list.
+     *
+     * The template type `Indicator` specifies the type of `aIndicator` object which is used to match against entries
+     * in the list. To check that an entry matches the given indicator, the `Matches()` method is invoked on each
+     * `Type` entry in the list. The `Matches()` method should be provided by `Type` class accordingly:
+     *
+     *     bool Type::Matches(const Indicator &aIndicator) const
+     *
+     * @param[in]  aIndicator  An indicator to match with entries in the list..
+     * @param[out] aPrevEntry  A pointer to output previous entry on success (when a match is found is the list).
+     *                         @p aPrevEntry is set to nullptr if the matching entry is head of the list. Otherwise it
+     *                         is updated to point to previous entry before the matching entry in the list.
+     *
+     * @retval OT_ERROR_NONE       A matching entry was found in the list and @p aPrevEntry was updated successfully.
+     * @retval OT_ERROR_NOT_FOUND  No matching entry The entry was not found in the list.
```

Comment:
This method returns a pointer, not `otError`.

Reply:
Thanks. Changed it.